File: 005_arbolAVL.py
#Implementar la estructura de datos: Árbol AVL.
import sys
import numpy as np

class Nodo:
    def __init__(self, elemento): #Método constructor recibe el elemento a insertar.
        self.elemento = elemento
        self.izquierdo = None
        self.derecho = None
        self.altura = 1


class arbolAVL: #Objeto AVL que contiene las propiedades y métodos.
    def __init__(self):
        self.raiz = None 
    
    
    def insertar(self, elemento): #Insertar el elemento.
        self.raiz = self._insertar(elemento, self.raiz)

    def _insertar(self, elemento, nodoActual):

        if nodoActual is None: #Si el nodo esta vacio insertamos el nuevo elemento.            
            return Nodo(elemento)

        elif elemento  < nodoActual.elemento: #Si el elemento es menor al elemento de la raiz ir a la izquierda del árbol.
            nodoActual.izquierdo = self._insertar(elemento, nodoActual.izquierdo)            
        else: #De otra manera nos vamos al lado derecho de árbol.
            nodoActual.derecho = self._insertar(elemento, nodoActual.derecho)                    
    
        #Actualizar la altura del nodo que se ha isertado.        
        nodoActual.altura = 1 + max(self.obtenerAltura(nodoActual.izquierdo), self.obtenerAltura(nodoActual.derecho))
        
        # Actualizar el balance del árbol y cumplir las propiedades del árbol.
        factorDeBalance = self.obtenerBalance(nodoActual)
                
        if factorDeBalance > 1: #Si el balance es mayor a 1.
            if elemento < nodoActual.izquierdo.elemento: #Si el elemento es menor al elemento del nodo izquierdo.
                return self.rotarAlaDerecha(nodoActual) #Hay que rotar a la derecha.
            else: #La otra manera es que el elemento es mayor al nodo izquierdo.
                nodoActual.izquierdo = self.rotarAlaIzquierda(nodoActual.izquierdo) #Hay que rotar a la izquierda primero.
                return self.rotarAlaDerecha(nodoActual) #Hay que rotar a la derecha después.

        if factorDeBalance < -1:
            if elemento > nodoActual.derecho.elemento:
                return self.rotarAlaIzquierda(nodoActual)
            else:
                nodoActual.derecho = self.rotarAlaDerecha(nodoActual.derecho)
                return self.rotarAlaIzquierda(nodoActual)

        return nodoActual

    """
    def remover(self, elemento):
        if self.raiz == None: 
            return #Si la raiz esta vacia salir del programa.
        else: #De otra manera hay que recorrer el árbol para cumplir con las propiedades del árbol AVL.
            self._remover(elemento, self.raiz)

    def _remover(self, elemento, nodoActual):        
        
        if not nodoActual:
            return nodoActual  #Si el nodo esta vacío hay que devolver ese nodo.

        if elemento < nodoActual.elemento:
            nodoActual.izquierdo = self._remover(nodoActual.izquierdo, elemento)
            return nodoActual #Devolver el nodo que tiene el elemento que vamos a eliminar.
        elif elemento > nodoActual.elemento:
            nodoActual.derecho = self._remover(nodoActual.derecho, elemento)
            return nodoActual #Devolver el nodo que tiene el elemento que vamos a eliminar.
        else:
            
            if nodoActual.izquierdo is None:
                nodoTemporal = nodoActual.derecho
                nodoActual = None
                return nodoTemporal
            elif nodoActual.derecho is None:
                nodoTemporal = nodoActual.izquierdo
                nodoActual = None
                return nodoTemporal
            
            nodoTemporal = self.obtenerPredecesor(nodoActual.derecho)
            nodoActual.elemento = nodoTemporal.elemento
            nodoActual.derecho = self._remover(nodoActual.derecho, nodoTemporal.elemento)
        
        if nodoActual is None:
            return nodoActual

        #Actualizar la altura del nodo que se ha isertado.
        nodoActual.altura = 1 + max(self.obtenerAltura(nodoActual.izquierdo), self.obtenerAltura(nodoActual.derecho))

        # Actualizar el balance del árbol y cumplir las propiedades del árbol.
        factorDeBalance = self.obtenerBalance(nodoActual)
        
        if factorDeBalance > 1: #Si el balance es mayor a 1.
            if elemento < nodoActual.izquierdo.elemento: #Si el elemento es menor al elemento del nodo izquierdo.
                return self.rotarAlaDerecha(nodoActual) #Hay que rotar a la derecha.
            else: #La otra manera es que el elemento es mayor al nodo izquierdo.
                nodoActual.izquierdo = self.rotarAlaIzquierda(nodoActual.izquierdo) #Hay que rotar a la izquierda primero.
                return self.rotarAlaDerecha(nodoActual) #Hay que rotar a la derecha después.

        if factorDeBalance < -1:
            if elemento > nodoActual.derecho.elemento:
                return self.leftRotate(nodoActual)
            else:
                nodoActual.derecho = self.rightRotate(nodoActual.derecho)
                return self.leftRotate(nodoActual)        
        
        return nodoActual
   
    """
    def rotarAlaIzquierda(self, z):
        y = z.derecho
        T2 = y.izquierdo
        y.izquierdo = z
        z.derecho = T2
        z.altura = 1 + max(self.obtenerAltura(z.izquierdo), self.obtenerAltura(z.derecho))
        y.altura = 1 + max(self.obtenerAltura(y.izquierdo), self.obtenerAltura(y.derecho))
        return y


    def rotarAlaDerecha(self, z):
        y = z.izquierdo
        T3 = y.derecho
        y.derecho = z
        z.izquierdo = T3
        z.altura = 1 + max(self.obtenerAltura(z.izquierdo), self.obtenerAltura(z.derecho))
        y.altura = 1 + max(self.obtenerAltura(y.izquierdo), self.obtenerAltura(y.derecho))
        return y

    # Devolver la altura de cada nodo.
    def obtenerAltura(self, nodoActual):
        if not nodoActual:
            return 0
        return nodoActual.altura

    # Devolver el balance del árbol.
    def obtenerBalance(self, nodoActual):
        if not nodoActual:
            return 0
        return self.obtenerAltura(nodoActual.izquierdo) - self.obtenerAltura(nodoActual.derecho)

    def obtenerPredecesor(self,nodoActual): #Al escojer el nodo derecho hay que buscar el valor minimo en el lado izquierdo del arbol.
        if nodoActual.izquierdo:            
            return self.obtenerPredecesor(nodoActual.izquierdo)
        return nodoActual


    def imprimirEnOrden(self):
        if self.raiz is None:
            return 
        else:
            self._imprimirEnOrden(self.raiz)

    def _imprimirEnOrden(self, nodoActual):
        if nodoActual is None:
            return      
        print("Imprimiendo en orden: ", end="")
        print(nodoActual.elemento)
        self._imprimirEnOrden(nodoActual.izquierdo)             
        self._imprimirEnOrden(nodoActual.derecho)

    def imprimirEnPostOrden(self):
        if self.raiz is None:
            return 
        else:
            self._imprimirEnPostOrden(self.raiz)

    def _imprimirEnPostOrden(self, nodoActual):
        if nodoActual is None:
            return              
        self._imprimirEnPostOrden(nodoActual.izquierdo)             
        self._imprimirEnPostOrden(nodoActual.derecho)
        print("Imprimiendo en post orden: ", end="")
        print(nodoActual.elemento)

    def imprimirEnPreOrden(self):
        if self.raiz is None:
            return 
        else:
            self._imprimirEnPreOrden(self.raiz)

    def _imprimirEnPreOrden(self, nodoActual):
        if nodoActual is None:
            return              
        self._imprimirEnPreOrden(nodoActual.izquierdo)              
        print("Imprimiendo en pre orden: ", end="")
        print(nodoActual.elemento)  
        self._imprimirEnPreOrden(nodoActual.derecho)

    def valorMinimo(self):
        if self.raiz is None:
            return
        else:
            self._valorMinimo(self.raiz)

    def _valorMinimo(self, nodoActual):
        if nodoActual.izquierdo:
            return self._valorMinimo(nodoActual.izquierdo)
        print(nodoActual.elemento)
        

    def valorMaximo(self):
        if self.raiz is None:
            return
        else:
            self._valorMaximo(self.raiz)

    def _valorMaximo(self, nodoActual):
        if nodoActual.derecho:
            return self._valorMaximo(nodoActual.derecho)
        print(nodoActual.elemento)
   

if __name__ == "__main__":
    #vector = np.random.randint(100,size=6)
    vector = [10,20,30,40]
    avl = arbolAVL()
    for i in  vector:
        avl.insertar(i)
    
    avl.imprimirEnOrden()
    print(" ")
    avl.imprimirEnPreOrden()
    print(" ")
    avl.imprimirEnPostOrden()
    print(" ")
    print("El valor máximo es:", avl.valorMaximo() )
    print("El valor mínimo es:", avl.valorMinimo() )
    #avl.remover(40)

File: 005_arbolAVL.js
//Implementar la estructura de datos: Árbol AVL.

class Nodo{
    constuctor(elemento){ //método constructor recibe el elemento a insertar.
        this.elemento = elemento
        this.izquierdo = null
        this.derecho = null
        this.altura = 1
    }
}


class arbolAVL{ //Objeto AVL que contiene las propiedades y métodos.
    constructor(){
        this.raiz == null     
    }
    
    insertar(elemento){ //Insertar el elemento.
        this.raiz = this._insertar(elemento, this.raiz)
    }

    _insertar(elemento, nodoActual){

        if (nodoActual == null){ //Si el nodo esta vacio insertamos el nuevo elemento.            
            return  new Nodo(elemento)
        }

        else if (elemento  < nodoActual.elemento){ //Si el elemento es menor al elemento de la raiz ir a la izquierda del árbol.
            nodoActual.izquierdo = this._insertar(elemento, nodoActual.izquierdo)
        }else{ //De otra manera nos vamos al lado derecho de árbol.
            nodoActual.derecho = this._insertar(elemento, nodoActual.derecho)
        }
    
        //Actualizar la altura del nodo que se ha isertado.        
        nodoActual.altura = 1 + max(this.obtenerAltura(nodoActual.izquierdo), this.obtenerAltura(nodoActual.derecho))
        
        // Actualizar el balance del árbol y cumplir las propiedades del árbol.
        let factorDeBalance = this.obtenerBalance(nodoActual)
                
        if (factorDeBalance > 1){ //Si el balance es mayor a 1.
            if (elemento < nodoActual.izquierdo.elemento){ //Si el elemento es menor al elemento del nodo izquierdo.
                return this.rotarAlaDerecha(nodoActual) //Hay que rotar a la derecha.
            }else{ //La otra manera es que el elemento es mayor al nodo izquierdo.
                nodoActual.izquierdo = this.rotarAlaIzquierda(nodoActual.izquierdo) //Hay que rotar a la izquierda primero.
                return this.rotarAlaDerecha(nodoActual) //Hay que rotar a la derecha después.
                }
            }

        if (factorDeBalance < -1){
            if (elemento > nodoActual.derecho.elemento){
                return this.rotarAlaIzquierda(nodoActual)
            }else{
                nodoActual.derecho = this.rotarAlaDerecha(nodoActual.derecho)
                return this.rotarAlaIzquierda(nodoActual)
                }
            }

        return nodoActual
    }
    
    rotarAlaIzquierda(z){
        y = z.derecho
        T2 = y.izquierdo
        y.izquierdo = z
        z.derecho = T2
        z.altura = 1 + max(this.obtenerAltura(z.izquierdo), this.obtenerAltura(z.derecho))
        y.altura = 1 + max(this.obtenerAltura(y.izquierdo), this.obtenerAltura(y.derecho))
        return y
    }


    rotarAlaDerecha(z){
        y = z.izquierdo
        T3 = y.derecho
        y.derecho = z
        z.izquierdo = T3
        z.altura = 1 + max(this.obtenerAltura(z.izquierdo), this.obtenerAltura(z.derecho))
        y.altura = 1 + max(this.obtenerAltura(y.izquierdo), this.obtenerAltura(y.derecho))
        return y
    }

    //Devolver la altura de cada nodo.
    obtenerAltura(nodoActual){
        if (nodoActual == null){
            return 0
        }
        return nodoActual.altura
    }

    // Devolver el balance del árbol.
    obtenerBalance(nodoActual){
        if (nodoActual == null){
            return 0
        }
        return this.obtenerAltura(nodoActual.izquierdo) - this.obtenerAltura(nodoActual.derecho)
    }

    imprimirEnOrden(){
        if (this.raiz == null){
            return 
        }else{
            this._imprimirEnOrden(this.raiz)
        }
    }

    _imprimirEnOrden(nodoActual){
        if (nodoActual == null){
            return
        }
        console.log(nodoActual.elemento)
        this._imprimirEnOrden(nodoActual.izquierdo)             
        this._imprimirEnOrden(nodoActual.derecho)
    }

} // cierra la clase AVL 

// Visualizar resultados en el navegador. 
let avl 
function setup() {
  createCanvas(800, 400) // Definir las dimensiones de la pantalla.
  avl = new arbolAVL()
  
  for (var i = 0; i < 10; i++) {
    avl.insertar(floor(random(0, 100)))
  }

  avl.imprimirEnOrden()
}

function draw() {
  background(0) // El color de fondo es negro.
  stroke(255) // El grosor del trazo.
  noFill() // La figura no tiene relleno.
  //avl.mostrarArbol()
}
